File: src/archives.cc
#include <nan.h>
#include "archives.h"
#include "lib.h"

using namespace Nan;

NAN_METHOD(ExtractContainer) {
    PWChar name = (PWChar) node::Buffer::Data(info[0]->ToObject());
    PWChar destination = (PWChar) node::Buffer::Data(info[1]->ToObject());
    WordBool replace = (WordBool) info[2]->BooleanValue();
    WordBool success = xelib.functions.ExtractContainer(name, path, replace);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(ExtractFile) {
    PWChar name = (PWChar) node::Buffer::Data(info[0]->ToObject());
    PWChar source = (PWChar) node::Buffer::Data(info[1]->ToObject());
    PWChar destination = (PWChar) node::Buffer::Data(info[2]->ToObject());
    WordBool success = xelib.functions.ExtractFile(name, source, destination);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetContainerFiles) {
    PWChar name = (PWChar) node::Buffer::Data(info[0]->ToObject());
    PInteger len = (PInteger) node::Buffer::Data(info[1]->ToObject());
    WordBool success = xelib.functions.GetContainerFiles(name, len);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetLoadedContainers) {
    PInteger len = (PInteger) node::Buffer::Data(info[0]->ToObject());
    WordBool success = xelib.functions.GetLoadedContainers(len);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(LoadContainer) {
    PWChar filePath = (PWChar) node::Buffer::Data(info[0]->ToObject());
    WordBool success = xelib.functions.LoadContainer(filePath);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

File: lib/archives.js
module.exports = function(lib, xelib, helpers) {
    let { Fail, GetString, GetStringArray, wcb } = helpers;

    // FILE HANDLING METHODS
    Object.assign(xelib, {
        ExtractContainer: function(name, dst, replace) {
            return lib.ExtractContainer(wcb(name), wcb(dst), replace);
        },
        ExtractFile: function(name, src, dst) {
            return lib.ExtractFile(wcb(name), wcb(src), wcb(dst));
        },
        GetContainerFiles: function(name, folder) {
            return GetStringArray(function(_len) {
                if (!lib.GetContainerFiles(wcb(name), wcb(folder), _len))
                    Fail(`Failed to get files in container ${name}`);
            });
        },
        GetFileContainer: function(filePath) {
            return GetString(function(_len) {
                if (!lib.GetFileContainer(wcb(filePath), _len))
                    Fail(`Failed to get file container for ${filePath}`);
            });
        },
        GetLoadedContainers: function() {
            return GetStringArray(function(_len) {
                if (!lib.GetLoadedContainers(_len))
                    Fail(`Failed to get loaded containers`);
            });
        },
        LoadContainer: function(filePath) {
            return lib.LoadContainer(wcb(filePath));
        },
        BuildArchive: function(
            name,
            folder,
            filePaths,
            archiveType,
            compress = false,
            share = false,
            af = '',
            ff = ''
        ) {
            lib.BuildArchive(
                wcb(name),
                wcb(folder),
                wcb(filePaths),
                archiveType,
                compress,
                share,
                af,
                ff
            ) || Fail(`Failed to build archive ${name}.`);
        }
    });
};
